platforms/allwinner-d1/core/src/drivers/sharp_display.rs

```diff
@@ -247,35 +208,64 @@ impl Draw {
     #[tracing::instrument(skip(self))]
     async fn draw_run(mut self) {
         loop {
-            let c = self.ctxt.lock().await;
-            // render into the buffer
-            let mut cmd = commands::WRITE_LINE;
-            if c.vcom {
-                cmd |= commands::VCOM_MASK;
-            }
+            let mut c = self.ctxt.lock().await;
+            self.buf.clear();
+
+            let mut drawn = 0;
+
+            // Are there ANY dirty lines?
+            if c.sdisp.dirty_line.iter().copied().any(identity) {
+                // render into the buffer
+                let mut cmd = commands::WRITE_LINE;
+                if c.vcom {
+                    cmd |= commands::VCOM_MASK;
+                }
+
+                // Write the command
+                let _ = self.buf.try_push(cmd);
+
+                let FullFrame { frame, dirty_line } = &mut c.sdisp;
 
-            // Write the command
-            self.buf.as_slice_mut()[FRAME_COMMAND_IDX] = cmd;
-
-            // Now we need to write all the lines, zip together the dest buffer
-            // with our current frame buffer
-            let out_lines =
-                self.buf.as_slice_mut()[FRAME_COMMAND_BYTES..].chunks_exact_mut(LINE_BYTES);
-            let in_lines = c.sdisp.frame.iter();
-            let lines = out_lines.zip(in_lines);
-
-            for (line, (oline, iline)) in &mut lines.enumerate() {
-                // Lines are 1-indexed on the wire
-                oline[LINE_COMMAND_IDX] = (line as u8) + 1;
-                // We keep our internal frame buffer in the same format as the wire
-                oline[LINE_COMMAND_BYTES..][..LINE_DATA_BYTES].copy_from_slice(iline);
-                // We don't need to write the dummy byte for the line
+                // Filter out to only the dirty lines, clearing the dirty flag
+                let all_lines = dirty_line.iter_mut().zip(frame.iter()).enumerate();
+                let dirty_lines = all_lines.filter_map(|(idx, (dirty, line))| {
+                    if *dirty {
+                        *dirty = false;
+                        Some((idx, line))
+                    } else {
+                        None
+                    }
+                });
+
+                // Now we need to write all the dirty lines, zip together the dest buffer
+                // with our current frame buffer
+                for (line, iline) in dirty_lines {
+                    // Lines are 1-indexed on the wire
+                    let _ = self.buf.try_push((line as u8) + 1);
+                    // We keep our internal frame buffer in the same format as the wire
+                    let _ = self.buf.try_extend_from_slice(iline);
+                    // dummy byte
+                    let _ = self.buf.try_push(0);
```

Comment:
nit/TIOLI: maybe the `try_push`/`try_extend_from_slice` calls should be `expect`ed? if the buffer ever doesn't have capacity we fucked up kinda bad, since it's preallocated to the max capacity we need...

Reply:
Yeah, probably. Too much embedded trauma with panic bloat makes me pretty wary of it though.